smart-contracts/contracts/UnlockTestV2.sol

```diff
@@ -0,0 +1,179 @@
+pragma solidity 0.4.24;
+
+/**
+ * @title The Unlock contract
+ * @author Julien Genestoux (unlock-protocol.com)
+ * This smart contract has 2 main roles:
+ *  1. Distribute discounts to discount token holders
+ *  2. Grant dicount tokens to users making referrals and/or publishers granting discounts.
+ * In order to achieve these 2 elements, it keeps track of several things such as
+ *  a. Deployed locks addresses and balances of discount tokens granted by each lock.
+ *  b. The total network product (sum of all key sales, net of discounts)
+ *  c. Total of discounts granted
+ *  d. Balances of discount tokens, including "frozen" tokens (which have been used to claim
+ * discounts and cannot be used/transfered for a given period)
+ *  e. Growth rate of Network Product
+ *  f. Growth rate of Discount tokens supply
+ * The smart contract has an owner who only can perform the following
+ *  - Upgrades
+ *  - Change in golden rules (20% of GDP available in discounts, and supply growth rate is at most
+ * 50% of GNP growth rate)
+ * NOTE: This smart contract is partially implemented for now until enough Locks are deployed and
+ * in the wild.
+ * The partial implementation includes the following features:
+ *  a. Keeping track of deployed locks
+ *  b. Keeping track of GNP
+ */
+
+import "openzeppelin-solidity/contracts/ownership/Ownable.sol";
+import "./Lock.sol";
+
+
+contract UnlockTestV2 is Ownable {
+
+  /**
+   * The struct for a lock
+   * We use deployed to keep track of deployments.
+   * This is required because both totalSales and yieldedDiscountTokens are 0 when initialized,
+   * which would be the same values when the lock is not set.
+   */
+  struct LockBalances {
+    bool deployed;
+    uint totalSales; // This is in wei
+    uint yieldedDiscountTokens;
+  }
+
+  modifier onlyFromDeployedLock() {
+    require(locks[msg.sender].deployed, "Only from previously deployed locks");
+    _;
+  }
+
+  uint public grossNetworkProduct;
+
+  uint public totalDiscountGranted;
+
+  // We keep track of deployed locks to ensure that callers are all deployed locks.
+  mapping (address => LockBalances) public locks;
+
+  // Events
+  event NewLock(
+    address indexed lockOwner,
+    address indexed newLockAddress
+  );
+
+  bool internal initialized;
+
+  function initialize(
```

Comment:
I think this is one of the changes compared to Unlock.sol, but I fail to see the other ones?

Reply:
This is actually on-par with Unlock now.  With the "Unstructured Storage Proxy" pattern, the only requirement for the contract itself is that instead of using a constructor, you use initialize.  

With this pattern, deploying v1 effectively creates 2 independent contracts.  You can interact with Unlock directly, as you had previously.  However if you want to upgrade in the future, you interact through the proxy.  When going through a proxy, data is stored in the proxy contract --- this means you get the logic from Unlock, including the layout of variables... but storage is completely independent.

Therefore we must change the constructor.  Anything assigned in a constructor would only impact Unlock directly... by using initialize, the same code can impact the proxy storage as well.

I'm going to submit a couple changes to clarify.  Give me a few..